utils/resolved_udp_conn_test.go

```diff
@@ -0,0 +1,298 @@
+	}
+
+	addr, err := net.ResolveUDPAddr("udp", mockServer.LocalAddr().String())
+	if err != nil {
+		mockServer.Close()
+		return nil, nil, err
+	}
+
+	conn, err := net.DialUDP("udp", nil, addr)
+	if err != nil {
+		mockServer.Close()
+		return nil, nil, err
+	}
+
+	return mockServer, conn, nil
+}
+
+func TestNewResolvedUDPConn(t *testing.T) {
+	hostPort := "blahblah:34322"
+
+	mockServer, clientConn, err := newUDPConn()
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	defer mockServer.Close()
+
+	resolver := mockResolver{}
+	resolver.
+		On("ResolveUDPAddr", "udp", hostPort).
+		Return(&net.UDPAddr{
+			IP:   net.IPv4(123, 123, 123, 123),
+			Port: 34322,
+		}, nil).
+		Once()
+
+	dialer := mockDialer{}
+	dialer.
+		On("DialUDP", "udp", (*net.UDPAddr)(nil), &net.UDPAddr{
+			IP:   net.IPv4(123, 123, 123, 123),
+			Port: 34322,
+		}).
+		Return(clientConn, nil).
+		Once()
+
+	conn, err := newResolvedUDPConn(hostPort, time.Hour, resolver.ResolveUDPAddr, dialer.DialUDP, log.NullLogger)
+	assert.NoError(t, err)
+	assert.NotNil(t, conn)
+
+	err = conn.Close()
+	assert.NoError(t, err)
+
+	// assert the actual connection was closed
+	assert.Error(t, clientConn.Close())
+
+	resolver.AssertExpectations(t)
+	dialer.AssertExpectations(t)
+}
+
+func TestResolvedUDPConnWrites(t *testing.T) {
+	hostPort := "blahblah:34322"
+
+	mockServer, clientConn, err := newUDPConn()
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	defer mockServer.Close()
+
+	resolver := mockResolver{}
+	resolver.
+		On("ResolveUDPAddr", "udp", hostPort).
+		Return(&net.UDPAddr{
+			IP:   net.IPv4(123, 123, 123, 123),
+			Port: 34322,
+		}, nil).
+		Once()
+
+	dialer := mockDialer{}
+	dialer.
+		On("DialUDP", "udp", (*net.UDPAddr)(nil), &net.UDPAddr{
+			IP:   net.IPv4(123, 123, 123, 123),
+			Port: 34322,
+		}).
+		Return(clientConn, nil).
+		Once()
+
+	conn, err := newResolvedUDPConn(hostPort, time.Hour, resolver.ResolveUDPAddr, dialer.DialUDP, log.NullLogger)
+	assert.NoError(t, err)
+	assert.NotNil(t, conn)
+
+	expectedString := "yo this is a test"
+	_, err = conn.Write([]byte(expectedString))
+	assert.NoError(t, err)
+
+	var buf = make([]byte, len(expectedString))
+	err = mockServer.SetReadDeadline(time.Now().Add(time.Second))
+	assert.NoError(t, err)
+	_, _, err = mockServer.ReadFrom(buf)
+	assert.NoError(t, err)
+	assert.Equal(t, []byte(expectedString), buf)
+
+	err = conn.Close()
+	assert.NoError(t, err)
+
+	// assert the actual connection was closed
+	assert.Error(t, clientConn.Close())
+
+	resolver.AssertExpectations(t)
+	dialer.AssertExpectations(t)
+}
+
+func TestResolvedUDPConnEventuallyDials(t *testing.T) {
+	hostPort := "blahblah:34322"
+
+	mockServer, clientConn, err := newUDPConn()
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	defer mockServer.Close()
+
+	resolver := mockResolver{}
+	resolveCall := resolver.
+		On("ResolveUDPAddr", "udp", hostPort).
+		Return(nil, fmt.Errorf("failed to resolve"))
+
+	timer := time.AfterFunc(time.Millisecond*100, func() {
```

Comment:
hm, this could create flaky test in CI due to short time interval. Could we use something more deterministic instead, like fail the resolve on connection creation, and then do a check-sleep loop until the connection is established from the background timer loop?
e.g.
```
for i:=0; i < 100; i++ {
    lock
    connected := (conn.conn != nil)
    unlock
    if connected { break }
    time.Sleep(10ms)
}
lock; defer unlock
assert.NotNil(t, conn.conn)
```